### ISSUE TITLE
Add kubernetes api updates to kubernetes manager

### ISSUE THREAD

**Issue by RmStorm** [renovatebot/renovate#5972]
**What would you like Renovate to be able to do?**
As described [here](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/) the kubernetes api will continue to change over time. When this happens the kubernetes team publishes which api updates should be executed by everyone. Currently I use renovatebot to update the docker images in my kubernetes manifests. It would be very helpful if it also checks the `apiversion` field and updates it if the api is being deprecated.

**Describe the solution you'd like**
Renovatebot should automatically update deprecated api's in the apiversion field in kubernetes manifests as described [here](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/).

**Describe alternatives you've considered**
Doing it by hand. It's fine but not amazing.

**Comment by rarkins** [comment]
I don't understand exactly what you need - an example/diff would help.

> if the api is being deprecated

e.g. What logic can Renovate use to determine if an API is being deprecated? Do you mean you don't want Renovate to update the API version if it's updated but *not* being deprecated? If you just mean that Renovate should update the `apiVersion` field whenever it changes, then please clarify that.

We would also need to identify the datasource we can use to know valid versions of the api version.

In summary, this may require:
- [ ] A new datasource used to fetch available versions of the API
- [ ] A new versioning scheme if Kubernetes are not using something we already support like SemVer (I noticed mentions of versions like `v1beta1` so that seems to indicate it's proprietary)
- [ ] The manager/kubernetes/extract code to detect/extract the version so that it can be updated

**Comment by RmStorm** [comment]
I think these depreciation cycles will happen maybe yearly for the next couple of years and everyone will need to update their kubernetes manifests. My suggestion is basically that a mapping is hardcoded into renovatebot that is updated when new depreciations are announced by kubernetes. So as an example I might have a DaemonSet defined somewhere in my repo. Currently only the image is checked and updated if a newer version is available.
``` yaml
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: prometheus-node-exporter
  namespace: monitoring
  labels:
    app: prometheus
    component: node-exporter
spec:
  template:
    metadata:
      name: prometheus-node-exporter
      labels:
        app: prometheus
        component: node-exporter
    spec:
      containers:
      - image: prom/node-exporter:v0.18.1
        name: prometheus-node-exporter
```
What I propose is that someone (maybe me?) updates a mapping when new depreciations are announced and the mapping would look like this:
``` javascript
const DepreciationMapping = {
  NetworkPolicy: {
    'extensions/v1beta1': 'networking.k8s.io/v1',
  },
  PodSecurityPolicy: {
    'extensions/v1beta1': 'policy/v1beta1',
  },
  DaemonSet: {
    'extensions/v1beta1': 'apps/v1',
    'apps/v1beta2': 'apps/v1',
  },
};
```
Now that I actually write it out I think the process might become to error prone since api depreciations also require additional actions beyond just updating the api version. For example certain fields might no longer be available in the new api.

I do think this is a problem faced by everyone using kubernetes and renovatebot could assist in smoothing this update process. I'm not sure what the best way to go about it is though.

For completeness the above kubernetes object should be updated like so:
``` yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: prometheus-node-exporter
  namespace: monitoring
  labels:
    app: prometheus
    component: node-exporter
spec:
  template:
    metadata:
      name: prometheus-node-exporter
      labels:
        app: prometheus
        component: node-exporter
    spec:
      containers:
      - image: prom/node-exporter:v0.18.1
        name: prometheus-node-exporter
```

**Comment by JamieMagee** [comment]
Sounds like #5558 could achieve this. Currently that PR is on hold until the `autoReplace` work is complete.

**Comment by rarkins** [comment]
@JamieMagee are you sure it's the right fit? Do you consider this is a dependency *name* change?

**Comment by RmStorm** [comment]
I've read through #5558 and I think I agree with @JamieMagee that this would be a good fit. I'm sorry for not finding that issue before making this one!

However I think the kubernetes depreciation cycle is more difficult then a straightforward rename since it can change fields that used to be optional into obligatory fields. In that case a `DaemonSet` or `Deployment` for example would have to be manually fixed up by a user.

**Comment by RmStorm** [comment]
Then again, that is actually no different than a major version bump in any dependency. You always have to go in and check that your stuff still works (even if you hope your automated tests catch issues). A replacement for a depreciated dependency should surely count as a major version bump so maybe this is not as big a problem as I was thinking earlier.

**Comment by rarkins** [comment]
Yes, I agree - it's not Renovate's job to create perfect PRs when there are breaking changes. Major version bumps of npm dependencies being an example. However in this case you consider it to be a *dependency name* bump and not a *dependency version* bump?

**Comment by RmStorm** [comment]
I would consider it to be a dependency version bump. Since changing an api version in a kubernetes manifest is like updating the version of a dependency of sorts.

**Comment by JamieMagee** [comment]
@rarkins Reading back through the thread, I'm not sure #5558 is the right match, and a hardcoded `k8s` datasource would be a better fit.

@RmStorm if I were to refactor your `DeprecationMapping` into how renovate would represent it internally, it would look something like:

```json
{
  "depName": "NetworkPolicy",
  "currentValue": "extensions/v1beta1",
  "updates": [
    {
      "newValue": "networking.k8s.io/v1"
    }
  ]
}
```

I think this is pretty straightforward to implement actually :smile:

**Comment by RmStorm** [comment]
@JamieMagee, I think your depreciation mapping is a lot more readable.

I wouldn't really know where to begin implementing it and I don't have any time next week. However, the week after I could make a start on a PR if you give me a couple pointers on how to go about it?

**Comment by JamieMagee** [comment]
Don't worry, it's not as bad as it looks. Either I can take a look, or I'll be able to walk you through it.

**Comment by HonkingGoose** [comment]
I saw that @JamieMagee made PR #5558 that might help with creating the necessary stuff to start work on this issue.
But PR #5558 is still in a draft state and is badly out-of-date compared to current `master`.

Are we `status:blocked` until PR #5558 lands? Or do we need some other thing before we can start? Or do we need to think some more about a possible solution?

**Comment by JamieMagee** [comment]
I would consider this blocked for now, due to #5558, unless someone comes up with an alternative approach.

**Comment by JamieMagee** [comment]
Not sure if #5558 is the right fit. Will need to do a small investigation and see if a hardcoded mapping, or autoreplacement is better suited.

**Comment by JamieMagee** [comment]
After investigation, I don't believe autoreplacements is the right way to do this, and a new `kubernetes-api` datasource would be more appropriate. Autoreplacements are more designed to replace the dependency name (`kind` in Kubernetes) not the version (`apiVersion` in Kubernetes).

The Kubernetes deprecated API migration guide[^1] should be the source of truth for the new datasource.

[^1]: https://kubernetes.io/docs/reference/using-api/deprecation-guide/

**Comment by rarkins** [comment]
@JamieMagee what's left to do now?

**Comment by JamieMagee** [comment]
@rarkins I have the manager and versioning, all that's left is the datasource to tie them together. It will be a static datasource generated from the Kubernetes [deprecated API migration guide][1]. Unfortunately, there's no machine readable format of these so it will have to be manually maintained.

[1]: https://kubernetes.io/docs/reference/using-api/deprecation-guide/

**Comment by JamieMagee** [comment]
This should now be complete in version [`32.221.0`](https://github.com/renovatebot/renovate/releases/tag/32.221.0)
